### ISSUE TITLE
Upgrading to v0.1.7 generates warning messages from the angular-cli compiler

### ISSUE THREAD

**Issue by ZiadJ** [typeorm/typeorm#1290]
The warning messages ended with this:
```
...
[C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\node_modules]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\node_modules]
[C:\Users\dev\Downloads\ng\node_modules]
[C:\Users\dev\Downloads\node_modules]
[C:\Users\dev\node_modules]
[C:\Users\node_modules]
[C:\node_modules]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\pg-query-stream]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\pg-query-stream.ts]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\pg-query-stream.js]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\pg-query-stream]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\pg-query-stream]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\pg-query-stream.ts]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\pg-query-stream.js]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\pg-query-stream]
[C:\Users\dev\Downloads\ng\rbecore\src\pg-query-stream]
[C:\Users\dev\Downloads\ng\rbecore\src\pg-query-stream.ts]
[C:\Users\dev\Downloads\ng\rbecore\src\pg-query-stream.js]
[C:\Users\dev\Downloads\ng\rbecore\src\pg-query-stream]
[C:\Users\dev\Downloads\ng\rbecore\src\pg-query-stream]
[C:\Users\dev\Downloads\ng\rbecore\src\pg-query-stream.ts]
[C:\Users\dev\Downloads\ng\rbecore\src\pg-query-stream.js]
[C:\Users\dev\Downloads\ng\rbecore\src\pg-query-stream]
 @ ./node_modules/typeorm/platform/PlatformTools.js 95:27-53
 @ ./node_modules/typeorm/entity-manager/MongoEntityManager.js
 @ ./node_modules/typeorm/index.js
 @ ./src/app/models/issue.ts
 @ ./src/app/components/vote-page/vote-page.component.ts
 @ ./src/app/app.module.ts
 @ ./src/main.ts
 @ multi webpack-dev-server/client?http://0.0.0.0:0 ./src/main.ts

WARNING in ./node_modules/typeorm/platform/PlatformTools.js
Module not found: Error: Can't resolve 'redis' in 'C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\platform'
resolve 'redis' in 'C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\platform'
  Parsed request is a module
  using description file: C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\package.json (relative path: ./platform)
    Field 'browser' doesn't contain a valid alias configuration
  after using description file: C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\package.json (relative path: ./platform)
    resolve as module
      C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\platform\node_modules doesn't exist or is not a directory
      C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\node_modules doesn't exist or is not a directory
      C:\Users\dev\Downloads\ng\rbecore\node_modules\node_modules doesn't exist or is not a directory
      C:\Users\dev\Downloads\ng\node_modules doesn't exist or is not a directory
      C:\Users\dev\Downloads\node_modules doesn't exist or is not a directory
      C:\Users\dev\node_modules doesn't exist or is not a directory
      C:\Users\node_modules doesn't exist or is not a directory
      C:\node_modules doesn't exist or is not a directory
      looking for modules in C:\Users\dev\Downloads\ng\rbecore\node_modules
        using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./node_modules)
          Field 'browser' doesn't contain a valid alias configuration
        after using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./node_modules)
          using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./node_modules/redis)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\node_modules\redis doesn't exist
            .ts
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\node_modules\redis.ts doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\node_modules\redis.js doesn't exist
            as directory
              C:\Users\dev\Downloads\ng\rbecore\node_modules\redis doesn't exist
      looking for modules in C:\Users\dev\Downloads\ng\rbecore\node_modules
        using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./node_modules)
          Field 'browser' doesn't contain a valid alias configuration
        after using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./node_modules)
          using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./node_modules/redis)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\node_modules\redis doesn't exist
            .ts
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\node_modules\redis.ts doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\node_modules\redis.js doesn't exist
            as directory
              C:\Users\dev\Downloads\ng\rbecore\node_modules\redis doesn't exist
      looking for modules in C:\Users\dev\Downloads\ng\rbecore\src
        using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./src)
          Field 'browser' doesn't contain a valid alias configuration
        after using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./src)
          using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./src/redis)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\src\redis doesn't exist
            .ts
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\src\redis.ts doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\src\redis.js doesn't exist
            as directory
              C:\Users\dev\Downloads\ng\rbecore\src\redis doesn't exist
      looking for modules in C:\Users\dev\Downloads\ng\rbecore\src
        using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./src)
          Field 'browser' doesn't contain a valid alias configuration
        after using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./src)
          using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./src/redis)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\src\redis doesn't exist
            .ts
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\src\redis.ts doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\src\redis.js doesn't exist
            as directory
              C:\Users\dev\Downloads\ng\rbecore\src\redis doesn't exist
[C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\platform\node_modules]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\node_modules]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\node_modules]
[C:\Users\dev\Downloads\ng\node_modules]
[C:\Users\dev\Downloads\node_modules]
[C:\Users\dev\node_modules]
[C:\Users\node_modules]
[C:\node_modules]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\redis]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\redis.ts]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\redis.js]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\redis]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\redis]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\redis.ts]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\redis.js]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\redis]
[C:\Users\dev\Downloads\ng\rbecore\src\redis]
[C:\Users\dev\Downloads\ng\rbecore\src\redis.ts]
[C:\Users\dev\Downloads\ng\rbecore\src\redis.js]
[C:\Users\dev\Downloads\ng\rbecore\src\redis]
[C:\Users\dev\Downloads\ng\rbecore\src\redis]
[C:\Users\dev\Downloads\ng\rbecore\src\redis.ts]
[C:\Users\dev\Downloads\ng\rbecore\src\redis.js]
[C:\Users\dev\Downloads\ng\rbecore\src\redis]
 @ ./node_modules/typeorm/platform/PlatformTools.js 100:27-43
 @ ./node_modules/typeorm/entity-manager/MongoEntityManager.js
 @ ./node_modules/typeorm/index.js
 @ ./src/app/models/issue.ts
 @ ./src/app/components/vote-page/vote-page.component.ts
 @ ./src/app/app.module.ts
 @ ./src/main.ts
 @ multi webpack-dev-server/client?http://0.0.0.0:0 ./src/main.ts

WARNING in ./node_modules/typeorm/platform/PlatformTools.js
Module not found: Error: Can't resolve 'sqlite3' in 'C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\platform'
resolve 'sqlite3' in 'C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\platform'
  Parsed request is a module
  using description file: C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\package.json (relative path: ./platform)
    Field 'browser' doesn't contain a valid alias configuration
  after using description file: C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\package.json (relative path: ./platform)
    resolve as module
      C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\platform\node_modules doesn't exist or is not a directory
      C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\node_modules doesn't exist or is not a directory
      C:\Users\dev\Downloads\ng\rbecore\node_modules\node_modules doesn't exist or is not a directory
      C:\Users\dev\Downloads\ng\node_modules doesn't exist or is not a directory
      C:\Users\dev\Downloads\node_modules doesn't exist or is not a directory
      C:\Users\dev\node_modules doesn't exist or is not a directory
      C:\Users\node_modules doesn't exist or is not a directory
      C:\node_modules doesn't exist or is not a directory
      looking for modules in C:\Users\dev\Downloads\ng\rbecore\node_modules
        using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./node_modules)
          Field 'browser' doesn't contain a valid alias configuration
        after using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./node_modules)
          using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./node_modules/sqlite3)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\node_modules\sqlite3 doesn't exist
            .ts
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\node_modules\sqlite3.ts doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\node_modules\sqlite3.js doesn't exist
            as directory
              C:\Users\dev\Downloads\ng\rbecore\node_modules\sqlite3 doesn't exist
      looking for modules in C:\Users\dev\Downloads\ng\rbecore\node_modules
        using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./node_modules)
          Field 'browser' doesn't contain a valid alias configuration
        after using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./node_modules)
          using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./node_modules/sqlite3)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\node_modules\sqlite3 doesn't exist
            .ts
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\node_modules\sqlite3.ts doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\node_modules\sqlite3.js doesn't exist
            as directory
              C:\Users\dev\Downloads\ng\rbecore\node_modules\sqlite3 doesn't exist
      looking for modules in C:\Users\dev\Downloads\ng\rbecore\src
        using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./src)
          Field 'browser' doesn't contain a valid alias configuration
        after using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./src)
          using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./src/sqlite3)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\src\sqlite3 doesn't exist
            .ts
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\src\sqlite3.ts doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\src\sqlite3.js doesn't exist
            as directory
              C:\Users\dev\Downloads\ng\rbecore\src\sqlite3 doesn't exist
      looking for modules in C:\Users\dev\Downloads\ng\rbecore\src
        using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./src)
          Field 'browser' doesn't contain a valid alias configuration
        after using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./src)
          using description file: C:\Users\dev\Downloads\ng\rbecore\package.json (relative path: ./src/sqlite3)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\src\sqlite3 doesn't exist
            .ts
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\src\sqlite3.ts doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              C:\Users\dev\Downloads\ng\rbecore\src\sqlite3.js doesn't exist
            as directory
              C:\Users\dev\Downloads\ng\rbecore\src\sqlite3 doesn't exist
[C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\platform\node_modules]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\typeorm\node_modules]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\node_modules]
[C:\Users\dev\Downloads\ng\node_modules]
[C:\Users\dev\Downloads\node_modules]
[C:\Users\dev\node_modules]
[C:\Users\node_modules]
[C:\node_modules]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\sqlite3]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\sqlite3.ts]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\sqlite3.js]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\sqlite3]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\sqlite3]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\sqlite3.ts]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\sqlite3.js]
[C:\Users\dev\Downloads\ng\rbecore\node_modules\sqlite3]
[C:\Users\dev\Downloads\ng\rbecore\src\sqlite3]
[C:\Users\dev\Downloads\ng\rbecore\src\sqlite3.ts]
[C:\Users\dev\Downloads\ng\rbecore\src\sqlite3.js]
[C:\Users\dev\Downloads\ng\rbecore\src\sqlite3]
[C:\Users\dev\Downloads\ng\rbecore\src\sqlite3]
[C:\Users\dev\Downloads\ng\rbecore\src\sqlite3.ts]
[C:\Users\dev\Downloads\ng\rbecore\src\sqlite3.js]
[C:\Users\dev\Downloads\ng\rbecore\src\sqlite3]
 @ ./node_modules/typeorm/platform/PlatformTools.js 105:27-45
 @ ./node_modules/typeorm/entity-manager/MongoEntityManager.js
 @ ./node_modules/typeorm/index.js
 @ ./src/app/models/issue.ts
 @ ./src/app/components/vote-page/vote-page.component.ts
 @ ./src/app/app.module.ts
 @ ./src/main.ts
 @ multi webpack-dev-server/client?http://0.0.0.0:0 ./src/main.ts

webpack: Compiled with warnings.
```
This may be an isolated issue but I thought I'd mention it still. Anyway reverting back to v0.1.6 fixes it for now.

**Comment by daniel-lang** [comment]
Have you read the changes you need to make with 0.1.7? We made a change to how webpack builds typeorm. More information can be found [here](https://github.com/typeorm/ionic-example#typeorm--017) or in the changelog.

**Comment by ZiadJ** [comment]
It looks like those updates are targeted at the Ionic framework but I'm not even using it. Is it not possible to build non-ionic apps without having to care for what looks like a purely Ionic related stuff?

**Comment by daniel-lang** [comment]
I assumed, that angular-cli uses webpack to build your project (just like ionic does to build apps) and therefore pointed you to the upgrade steps for Ionic since there are the same for you if you use webpack (you need to add NormalModuleReplacementPlugin)
If you don't use webpack, than please tell me what tool/module loader you use.

**Comment by daniel-lang** [comment]
@ZiadJ did you manage to get it running?

**Comment by ZiadJ** [comment]
The project i'm on does use Webpack but i'm still stuck on v0.1.6 by lack of time to give it a try. But feel free to close the thread for now if you wish and I'll reopen a new thread later if the issues still show up later.

**Comment by daniel-lang** [comment]
Okay, I will close this issue then

**Comment by loki2302** [comment]
@daniel-lang I'm trying to make TypeORM work in an Angular 5 application and no luck so far. As far as I understand:

1. If I want to use TypeORM in the browser environment, I should use webpack and [this piece of configuration](https://github.com/typeorm/ionic-example#typeorm--017).
2. Angular team [says here](https://github.com/angular/angular-cli/issues/1656#issuecomment-239366723) that they are not going to allow overriding/extending the webpack configuration produced by their Angular CLI.

So basically there's no way to use Angular CLI and TypeORM at the same time.

**Comment by daniel-lang** [comment]
@loki2302 thank you for that information!
I didn't know angular cli hides its config. If you have an idea on how it could work, I'm all for it but I'm afraid the current setup of TypeORM forces a motification of the config

**Comment by jelling** [comment]
Re: the original issue:

`Module not found: Error: Can't resolve 'xyz' in '/somepath'`

I got a ton of these after sharing TypeORM models from the server to the client. I.e. I was importing the models into client-side components (isomorphic types).

The underlying problem seems to be that `ng build` follows the annotations in the models and subsequently tries to package a bunch of server-side node_modules. 

The fix that worked for me was adding this line to my tsconfig.app.json.

```  
"compilerOptions": {
     ...
    "paths": {
      "typeorm": ["../../server/node_modules/typeorm/typeorm-model-shim.js"]
    }
}
```
The path for "typeorm" will vary based on your apps setup. But the typeorm-model-shim somehow tells the builder to ignore the class annotations. `ng build --prod` then worked. 

Note: this was with Angular 6, which matters because people have been saying that TypeORM is not compatible with Angular 6 due to this issue. @daniel-lang perhaps the docs could be updated with this info?

**Comment by ZiadJ** [comment]
Thanks @jelling but, while that solved the build process issue but now I'm getting a blank page, I'm now getting a blank page so I'm not sure how to proceed now. The only way to get the page running like before is to remove any reference to the components that uses the typeorm entities but that beats the purpose of it.

**Comment by pleerock** [comment]
@ZiadJ one of the solutions to using models on the client is to override references to typeorm via webpack

**Comment by zoezydk** [comment]
> Re: the original issue:
> 
> `Module not found: Error: Can't resolve 'xyz' in '/somepath'`
> 
> I got a ton of these after sharing TypeORM models from the server to the client. I.e. I was importing the models into client-side components (isomorphic types).
> 
> The underlying problem seems to be that `ng build` follows the annotations in the models and subsequently tries to package a bunch of server-side node_modules.
> 
> The fix that worked for me was adding this line to my tsconfig.app.json.
> 
> ```
> "compilerOptions": {
>      ...
>     "paths": {
>       "typeorm": ["../../server/node_modules/typeorm/typeorm-model-shim.js"]
>     }
> }
> ```
> The path for "typeorm" will vary based on your apps setup. But the typeorm-model-shim somehow tells the builder to ignore the class annotations. `ng build --prod` then worked.
> 
> Note: this was with Angular 6, which matters because people have been saying that TypeORM is not compatible with Angular 6 due to this issue. @daniel-lang perhaps the docs could be updated with this info?


@jelling What if you aren't interested (yet) in doing a build but just want to use `ng serve` to test database connectivity in the browser and are faced with a similar list of errors? What is the solution here? Does the adding of the same path in tsconfig.app.json apply? Or is there a similar but different solution to the problem? Do you know? Thanks in advance for any feedback.

**Comment by ZiadJ** [comment]
@pleerock 

> EDIT: ah I see @jelling already provided a solution

As I already mentioned that's not a proper solution since it only allows the build to complete. Serving the app leads to a blank page.

**Comment by ZiadJ** [comment]
@pleerock In the other thread which you just locked you stated that _"TypeORM didn't any compatibility with Angular because its not related to Angular anyhow."_. But this thread testifies otherwise and I never said that TypeORM was related to Angular anyhow. My code was and is still running fine on v0.1.6 which is why I stated that. Everything broke right after and I've been patient for a year already hoping I'd be able resume work at some point but that never happened.

**Comment by jelling** [comment]
@ZiadJ the bug here is in your expectations. TypeORM is an open-source project, meaning no one is making money off of providing this software that makes your life easier and they're certainly not required to give you support on the timeline you dictate.

More importantly, the author is correct that TypeORM is not related or dependent on Angular. The issues related to Angular are because of Angular's build settings. I provided one solution but other issues may exist. Your white-screen issue could be almost anything - basically the rendering of the app is broken - but if it's related to TypeORM, it's how Angular chooses to handle TypeORM code, not the actual TypeORM.

**Comment by zoezydk** [comment]
@ZiadJ Since I'm as stuck as you are with this issue and since the "why-you-shouldn't-complain-but-be-happy-about-open-source" people are impervious to warranted critique aimed at improving the usability of this and other open-source software (which they really should if they were interested in improving it and making it available to as many people as possible which is the general idea) by their users, I'd recommend taking a look at Loopback. As it looks now, you have to either find a way of "hacking" the incompatibility between the Ng and typeorm CLI or go with another alternative. Since you get a blank screen when trying to serve the app, I'm not even interested in going there. I have spent way too much time on this as it is. To your question in the "conveniently" locked thread, I'd recommend to take a look at Loopback. You can find some good video-tutorials on the framework as well as that Loopback is building a neat API for you and looks quite promising. To anyone else, I'm sorry being the inexperienced dummy that I am to waste your time reading this and for asking questions of support in a support forum.

**Comment by haschu** [comment]
I came across the same problem and I just wanted to point out, that the [solution](https://github.com/typeorm/typeorm/issues/1290#issuecomment-420143054) provided by @jelling works just fine (with the latest version of both, TypeORM and Angular).

No "white screen" while serving the app or anything else. I guess the other problems mentioned in this thread aren't related to a TypeORM bug.

**Comment by jordancue** [comment]
@haschu when I follow this logic I'm getting a lot of `Circular dependency detected` warnings, and in my browser the error `Cannot access 'Organization' before initialization` (Organization is one of my entities) because it's being used in TypeORM decorators, and my entities have circular relations to one another. How were you able to solve that?

**Comment by haschu** [comment]
@jordancue I don’t have any of your issues so there was nothing for me to solve.

Would you mind to provide a minor repo or jsfiddle? Otherwise it’s hard for people to help you with your specific problem.

**Comment by jordancue** [comment]
I'm working on extracting out part to show the issue (it's a monorepo build on nx ), but the core of it is that the TypeORM entities are contained within an nx lib I generated in the hopes of combining my Angular and node repos and sharing the code that way. 

In the entities, I'm using relationships like:

**Organization.ts**
```
@OneToMany(type => OrganizationAlias, alias => alias.organization)
aliases: OrganizationAlias[];
```

**OrganizationAlias.ts**
```
@ManyToOne(type => Organization, organization => organization.aliases, { onDelete: 'CASCADE' })
@JoinColumn({ name: 'organizationId' })
organization: Organization;
```

Organization.ts imports OrganizationAlias.ts and vice versa -- this wasn't a problem on Node, but with Angular CLI it's giving me a bunch of problems. I've noticed that it's actually related to **the decorators** themselves in TypeORM... if I remove the decorators (and the class names inside them) I don't get cyclic dependencies even though the classes are still referenced and imported in the files, e.g. `aliases: OrganizationAlias[]`. I think this is related to the way metadata reflection is handled within decorators.

**Comment by jordancue** [comment]
FYI I've figured out the cause is https://github.com/typeorm/typeorm/issues/2059 but I can't figure out how to fix it

**Comment by haschu** [comment]
Ok, I’ll take a look when I’m at the office tomorrow :)

**Comment by haschu** [comment]
Though it would still be helpful if you could post a minimal repo of your code and cli setup.

**Comment by haschu** [comment]
@jordancue I checked your use-case and I got no cycle dependency issues.

**Comment by jordancue** [comment]
@haschu here you go: https://github.com/jordancue/cyclic-dep-ng-typeorm

**Comment by haschu** [comment]
@jordancue Ok, now I see your problem - but it's not a TypeORM issue, it's a problem within your code. You're creating in fact a circular dependencies because `User` and `Photo` are referencing each other and you're creating an instance of `User` in your angular code.

I see three options for you:

1. Don't create instances of your entities in your frontend, just work with the plain types:
```ts
@Injectable({
    providedIn: 'root'
})
export class UserService {
    // Not the most elegant solution, but you get the idea
    currentUser = {} as User;
    photos: Photo[];

    constructor() {

        this.currentUser.name = 'John Bonham';
        this.photos = Array<Photo>();
        
    }
}

```
I personally use my entities only via generics. For example you could write a `EntityFactory<T>` class.

2. Drop the `@ManyToOne` part from `Photo` if you don't need it. The many-to-one relation still works fine when you're dealing with `User` - and you don't have a circular reference any more.
So if you don't have to deal with photos independently from a user, you should be fine.

4. Use [EntitySchema](https://typeorm.io/#/separating-entity-definition)


Anyways, the solutions are all about avoiding circular dependencies, which would always cause warnings and unintended side effects, independent from TypeORM.

I hope that helps :)

**Comment by jordancue** [comment]
@haschu thanks so much for taking a look!

I actually did already start going down the road of using `EntitySchema` yesterday and am switching all of my entities over to that format. I wouldn't classify this as either an issue with my code or with TypeORM after looking into it, just an incompatibility.

The constructors actually aren't the issue, it's using the classes `Photo` and `User` in each other's decorators (in the `ManyToOne` and `OneToMany`) that triggers it, in addition to injecting a dependency of `UserService` into the `UserComponent`. I'm actually not sure why TypeORM requires the actual class inside those decorators instead of just a type like it does column definitions, but that's not compatible with Angular's dependency building out of the box (which TypeORM doesn't profess to support anyway).

**Comment by haschu** [comment]
You're welcome :)

The downside of using `EntitySchema` for me is, that it cannot be used in combination with class-validator...

Anyways, I'm currently researching if I could come up with a pull-request, that might prevent those issues while still being able to use the decorator-style approach.
